### PR TITLE
Support upto Federation v2.7 using graphene-directives

### PR DESCRIPTION
**Pull request by mak626**
Supported versions, v1.0, 2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7

All directives are purely based on [apollo-specs](https://specs.apollo.dev/#sec-All-Schemas).

Some non existent directive like`@extend` which acted like `@key` was deprecated.

Migration Guide
---------------

For V1 `@extend` can be replaced by `@extends` and `@key` combination
For V2 just `@key` is enough. 
Deprecate `enable_federation_2` in favour of `federation_version`

# Updated Docs
---------------------
...
------------------------
## Supported Features

* `sdl` (`_service` on field): enable to add schema in federation (as is)

## Apollo Spec Supported

- [x] v1.0
- [x] v2.0
- [x] v2.1
- [x] v2.2
- [x] v2.3
- [x] v2.4
- [x] v2.5 
- [x] v2.6 `STABLE_VERSION` . Rover dev supports only upto v2.6
- [x] v2.7 `LATEST_VERSION`

All directives could be easily integrated with the help of [graphene-directives](https://github.com/strollby/graphene-directives). 
Now every directive's values are validated at run time itself by [graphene-directives](https://github.com/strollby/graphene-directives).

### Directives (v2.7)

```graphql
directive @composeDirective(name: String!) repeatable on SCHEMA
directive @extends on OBJECT | INTERFACE
directive @external on OBJECT | FIELD_DEFINITION
directive @key(fields: FieldSet!, resolvable: Boolean = true) repeatable on OBJECT | INTERFACE
directive @inaccessible on
  | FIELD_DEFINITION
  | OBJECT
  | INTERFACE
  | UNION
  | ENUM
  | ENUM_VALUE
  | SCALAR
  | INPUT_OBJECT
  | INPUT_FIELD_DEFINITION
  | ARGUMENT_DEFINITION
directive @interfaceObject on OBJECT
directive @override(from: String!, label: String) on FIELD_DEFINITION
directive @provides(fields: FieldSet!) on FIELD_DEFINITION
directive @requires(fields: FieldSet!) on FIELD_DEFINITION
directive @shareable repeatable on FIELD_DEFINITION | OBJECT
directive @tag(name: String!) repeatable on
  | FIELD_DEFINITION
  | INTERFACE
  | OBJECT
  | UNION
  | ARGUMENT_DEFINITION
  | SCALAR
  | ENUM
  | ENUM_VALUE
  | INPUT_OBJECT
  | INPUT_FIELD_DEFINITION
directive @authenticated on
    FIELD_DEFINITION
  | OBJECT
  | INTERFACE
  | SCALAR
  | ENUM
directive @requiresScopes(scopes: [[federation__Scope!]!]!) on
    FIELD_DEFINITION
  | OBJECT
  | INTERFACE
  | SCALAR
  | ENUM
directive @policy(policies: [[federation__Policy!]!]!) on
  | FIELD_DEFINITION
  | OBJECT
  | INTERFACE
  | SCALAR
  | ENUM
scalar federation__Policy
scalar federation__Scope
scalar FieldSet
```

Read about directives in [official documentation](https://www.apollographql.com/docs/federation/federated-types/federated-directives)

Each type which is decorated with `@key` or `@extends` is added to the `_Entity` union.
The [`__resolve_reference` method](https://www.apollographql.com/docs/federation/api/apollo-federation/#__resolvereference) can be defined for each type that is an entity.
Note that since the notation with double underscores can be problematic in Python for model inheritance this resolver method can also be named `_resolve_reference` (the `__resolve_reference` method will take precedence if both are declared).

This method is called whenever an entity is requested as part of the fulfilling a query plan.
If not explicitly defined, the default resolver is used.
The default resolver just creates instance of type with passed fieldset as kwargs, see [`entity.get_entity_query`](graphene_federation/entity.py) for more details
* You should define `__resolve_reference`, if you need to extract object before passing it to fields resolvers (example: [FileNode](integration_tests/service_b/src/schema.py))
* You should not define `__resolve_reference`, if fields resolvers need only data passed in fieldset (example: [FunnyText](integration_tests/service_a/src/schema.py))
Read more in [official documentation](https://www.apollographql.com/docs/apollo-server/api/apollo-federation/#__resolvereference).

------------------------

## Example

Here is an example of implementation based on the [Apollo Federation introduction example](https://www.apollographql.com/docs/federation/).
It implements a federation schema for a basic e-commerce application over three services: accounts, products, reviews.

### Accounts
First add an account service that expose a `User` type that can then be referenced in other services by its `id` field:

```python
from graphene import Field, Int, ObjectType, String

from graphene_federation import build_schema, key


@key("id")
class User(ObjectType):
    id = Int(required=True)
    username = String(required=True)

    def __resolve_reference(self, info, **kwargs):
        """
        Here we resolve the reference of the user entity referenced by its `id` field.
        """
        return User(id=self.id, email=f"user_{self.id}@mail.com")


class Query(ObjectType):
    me = Field(User)


schema = build_schema(query=Query, enable_federation_2=True)
```

### Product
The product service exposes a `Product` type that can be used by other services via the `upc` field:

```python
from graphene import Argument, Int, List, ObjectType, String

from graphene_federation import build_schema, key


@key("upc")
class Product(ObjectType):
    upc = String(required=True)
    name = String(required=True)
    price = Int()

    def __resolve_reference(self, info, **kwargs):
        """
        Here we resolve the reference of the product entity referenced by its `upc` field.
        """
        return Product(upc=self.upc, name=f"product {self.upc}")


class Query(ObjectType):
    topProducts = List(Product, first=Argument(Int, default_value=5))


schema = build_schema(query=Query, enable_federation_2=True)
```

### Reviews
The reviews service exposes a `Review` type which has a link to both the `User` and `Product` types.
It also has the ability to provide the username of the `User`.
On top of that it adds to the `User`/`Product` types (that are both defined in other services) the ability to get their reviews.

```python
from graphene import Field, Int, List, ObjectType, String

from graphene_federation import build_schema, external, key, provides


@key("id")
class User(ObjectType):
    id = external(Int(required=True))
    reviews = List(lambda: Review)

    def resolve_reviews(self, info, *args, **kwargs):
        """
        Get all the reviews of a given user. (not implemented here)
        """
        return []


@key("upc")
class Product(ObjectType):
    upc = external(String(required=True))
    reviews = List(lambda: Review)


class Review(ObjectType):
    body = String()
    author = provides(Field(User), fields="username")
    product = Field(Product)


class Query(ObjectType):
    review = Field(Review)


schema = build_schema(query=Query, enable_federation_2=True)
```

### Federation

Note that each schema declaration for the services is a valid graphql schema (it only adds the `_Entity` and `_Service` types).
The best way to check that the decorator are set correctly is to request the service sdl:

```python
from graphql import graphql

query = """
query {
    _service {
        sdl
    }
}
"""

result = graphql(schema, query)
print(result.data["_service"]["sdl"])
```

Those can then be used in a federated schema.

You can find more examples in the unit / integration tests and [examples folder](examples/).

There is also a cool [example](https://github.com/preply/graphene-federation/issues/1) of integration with Mongoengine.

------------------------
### build_schema new arguments

- `schema_directives` (`Collection[SchemaDirective]`): Directives that can be defined at `DIRECTIVE_LOCATION.SCHEMA` with their argument values.
- `include_graphql_spec_directives` (`bool`): Includes directives defined by GraphQL spec (`@include`, `@skip`, `@deprecated`, `@specifiedBy`)
- `enable_federation_2` (`bool`): Whether to enable federation 2 directives (default False)
- `federation_version` (`FederationVersion`): Specify the version explicit (default LATEST_VERSION)

In case both enable_federation_2 and federation_version are specified, federation_version is given higher priority

### Directives Additional arguments

-  `federation_version`: (`FederationVersion` = `LATEST_VERSION`) : You can use this to take a directive from a particular federation version

Note: The `federation_version` in `build_schema` is given higher priority. If the directive you have chosen is not compatible, it will raise an error

### Custom Directives

You can define custom directives as follows

```python
from graphene import Field, ObjectType, String
from graphql import GraphQLArgument, GraphQLInt, GraphQLNonNull

from graphene_federation import DirectiveLocation, FederationDirective
from graphene_federation import build_schema

CacheDirective = FederationDirective(
    name="cache",
    locations=[DirectiveLocation.FIELD_DEFINITION, DirectiveLocation.OBJECT],
    args={
        "maxAge": GraphQLArgument(
            GraphQLNonNull(GraphQLInt), description="Specifies the maximum age for cache in seconds."
        ),
    },
    description="Caching directive to control cache behavior.",
    spec_url="https://specs.example.dev/directives/v1.0",
)

cache = CacheDirective.decorator()


@cache(max_age=20)
class Review(ObjectType):
    body = cache(field=String(),max_age=100)


class Query(ObjectType):
    review = Field(Review)


schema = build_schema(
    query=Query,
    directives=(CacheDirective,),
    enable_federation_2=True,
)
```

This will automatically add @link and @composeDirective to schema

```graphql
extend schema
	@link(url: "https://specs.apollo.dev/federation/v2.6", import: ["@composeDirective"])
	@link(url: "https://specs.example.dev/directives/v1.0", import: ["@cache"])
	@composeDirective(name: "@cache")

"""Caching directive to control cache behavior."""
directive @cache(
  """Specifies the maximum age for cache in seconds."""
  maxAge: Int!
) on FIELD_DEFINITION | OBJECT

type Query {
  review: Review
  _service: _Service!
}

type Review  @cache(maxAge: 20) {
  body: String @cache(maxAge: 100)
}
```

If you wish to add the schema_directives `@link` `@composeDirective` manually. 
You can pass the `add_to_schema_directives` as `False`

```python
from graphene import Field, ObjectType, String
from graphql import GraphQLArgument, GraphQLInt, GraphQLNonNull

from graphene_federation import DirectiveLocation, FederationDirective, build_schema, compose_directive, link_directive

CacheDirective = FederationDirective(
    name="cache",
    locations=[DirectiveLocation.FIELD_DEFINITION, DirectiveLocation.OBJECT],
    args={
        "maxAge": GraphQLArgument(
            GraphQLNonNull(GraphQLInt), description="Specifies the maximum age for cache in seconds."
        ),
    },
    description="Caching directive to control cache behavior.",
    add_to_schema_directives=False
)

cache = CacheDirective.decorator()


@cache(max_age=20)
class Review(ObjectType):
    body = cache(field=String(), max_age=100)


class Query(ObjectType):
    review = Field(Review)


schema = build_schema(
    query=Query,
    directives=(CacheDirective,),
    schema_directives=(
        link_directive(url="https://specs.example.dev/directives/v1.0", import_=['@cache']),
        compose_directive(name='@cache'),
    ),
    enable_federation_2=True,
)
```

### Custom field name

When using decorator on a field with custom name

####  Case 1 (auto_camelcase=False)

```python
@key("identifier")
@key("validEmail")
class User(ObjectType):
    identifier = ID()
    email = String(name="validEmail")

class Query(ObjectType):
    user = Field(User)

schema = build_schema(query=Query, enable_federation_2=True, auto_camelcase=False) # Disable auto_camelcase
```

This works correctly.
By default `fields` of `@key`,`@requires` and `@provides` are not converted to camel case if `auto_camelcase` is set to `False`

#### Case 2 (auto_camelcase=True)
```python
@key("identifier")
@key("valid_email")
class User(ObjectType):
    identifier = ID()
    email = String(name="valid_email")

class Query(ObjectType):
    user = Field(User)

schema = build_schema(query=Query, enable_federation_2=True) # auto_camelcase Enabled
```

This will raise an error `@key, field "validEmail" does not exist on type "User"`. 
Because The decorator auto camel-cased the `field` value of key, as schema has `auto_camelcase=True` (default)

To fix this, pass `auto_case=False` in the `@key`, `@requires` or `@provides` argument

```python
@key("identifier")
@key("valid_email", auto_case=False)
class User(ObjectType):
    identifier = ID()
    email = String(name="valid_email")

class Query(ObjectType):
    user = Field(User)

schema = build_schema(query=Query, enable_federation_2=True) # auto_camelcase=True
```

### CODE REVIEW

**Comment by arunsureshkumar** [comment]
x

> Supported versions, v1.0, 2.1, 2.2, 2.3
> 
> All directives are purely based on [apollo-specs](https://specs.apollo.dev/#sec-All-Schemas).
> 
> ## Sample
> ```python
> schema = build_schema(
>     query=Query,
>     mutation=Mutation,
>     types=(UserType,),
>     directives=(CacheDirective,),
>     schema_directives=(
>         link_directive(url="https://specs.example.dev/directives/v1.0", import_=[str(CacheDirective)]),
>         compose_directive(name=str(CacheDirective)),
>     ),
>     federation_version="2.5",
> )
> ```
> 
> Some non existent directive like`@extend` which acted like `@key` was deprecated.
> 
> ## Migration Guide
> For V1 `@extend` can be replaced by `@extends` and `@key` combination For V2 just `@key` is enough.
> 
> ## Latest Spec
> ```graphql
> directive @composeDirective(name: String!) repeatable on SCHEMA
> directive @extends on OBJECT | INTERFACE
> directive @external on OBJECT | FIELD_DEFINITION
> directive @key(fields: FieldSet!, resolvable: Boolean = true) repeatable on OBJECT | INTERFACE
> directive @inaccessible on
>   | FIELD_DEFINITION
>   | OBJECT
>   | INTERFACE
>   | UNION
>   | ENUM
>   | ENUM_VALUE
>   | SCALAR
>   | INPUT_OBJECT
>   | INPUT_FIELD_DEFINITION
>   | ARGUMENT_DEFINITION
> directive @interfaceObject on OBJECT
> directive @override(from: String!) on FIELD_DEFINITION
> directive @provides(fields: FieldSet!) on FIELD_DEFINITION
> directive @requires(fields: FieldSet!) on FIELD_DEFINITION
> directive @shareable repeatable on FIELD_DEFINITION | OBJECT
> directive @tag(name: String!) repeatable on
>   | FIELD_DEFINITION
>   | INTERFACE
>   | OBJECT
>   | UNION
>   | ARGUMENT_DEFINITION
>   | SCALAR
>   | ENUM
>   | ENUM_VALUE
>   | INPUT_OBJECT
>   | INPUT_FIELD_DEFINITION
> scalar FieldSet
> ```
> 
> ## Pending
> * [ ]  Fix tests
> * [ ]  Update documentation

@mak626  - 

1. Let's create an enum of versions supported. This will help the user to select the same without confusion about what we support currently.
2. Modify graphene-directive to add a spec URL, which should be optional, but make sure while using federation, it should use that while building supergraph.
3. compose_directive should be auto-populated; don't take this as an argument.
4. schema_directives=(
        link_directive(url="https://specs.example.dev/directives/v1.0", import_=[str(CacheDirective)]),
        compose_directive(name=str(CacheDirective)),
    ), - This should be based on the version enum.

**Comment by mak626** [comment]
@arunsureshkumar 

```python
from graphql import GraphQLArgument, GraphQLInt, GraphQLNonNull, GraphQLString
from graphene_federation import DirectiveLocation, FederationDirective

CacheDirective = FederationDirective(
    name="cache",
    locations=[DirectiveLocation.FIELD_DEFINITION, DirectiveLocation.OBJECT],
    args={
        "maxAge": GraphQLArgument(
            GraphQLNonNull(GraphQLInt), description="Specifies the maximum age for cache in seconds."
        ),
        "swr": GraphQLArgument(GraphQLInt, description="Stale-while-revalidate value in seconds. Optional."),
        "scope": GraphQLArgument(GraphQLString, description="Scope of the cache. Optional."),
    },
    description="Caching directive to control cache behavior of fields or fragments.",
    spec_url="https://specs.example.dev/directives/v1.0",
    # add_to_schema_directives=False , if given will not generate schema directives
)

cache = CacheDirective.decorator()

schema = build_schema(
    query=Query,
    mutation=Mutation,
    types=(UserExtendedType,),
    directives=(StellateCacheDirective,),
    federation_version=FederationVersion.VERSION_2_3,
    # Following still works, if user wants to give something custom
    # schema_directives=(
    #    link_directive(url="https://specs.example.dev/directives/v1.0", import_=["@auth"]),
    #    compose_directive(name="@auth"),
    # ),
)
```

- Graphene directive will not support spec_url because it is defined purely based on original [spec](https://github.com/graphql-python/graphql-core/blob/main/src/graphql/type/directives.py). Instead we have class `FederationDirective`

**Review by abhinand-c — COMMENTED** [review]
Could we add docstring to the functions, and modules to clarify intentions.

File: graphene_federation/directives/shareable.py
Comment (abhinand-c) [inline graphene_federation/directives/shareable.py]: Use of .. Multi level relative imports, should be discouraged in favour of absolute imports

**Review by abhinand-c — COMMENTED** [review]
On a quick glance looks good, would definitely require in-depth review to iron out issues,  and anti-patterns

File: README.md
Comment (abhinand-c) [inline README.md]: Use **ComposableDirective** as name, or something similar.

**Comment by mak626** [comment]
@abhinand-c should we deprecate use of `enable_federation2` in build_schema

**Comment by abhinand-c** [comment]
Yes, we can deprecate it in favour of federation version

**Comment by abhinand-c** [comment]
Since, this release introduces significant changes, we can remove enable_federation and mention breaking in the docs.

File: graphene_federation/apollo_versions/v2_2.py
Comment (noelleleigh) [inline graphene_federation/apollo_versions/v2_2.py]: This `@shareable` directive isn't being updated with the new version here. This change fixes that:

```suggestion
        {directive.name: directive for directive in [shareable_directive]}
```

File: graphene_federation/schema.py
Comment (noelleleigh) [inline graphene_federation/schema.py]: In a multi-thread environment, this function could be called multiple times. If `PageInfo` is already marked as `@shareable`, then an exception is raised:

<details>

<summary>Large stack trace from Django startup</summary>

```
Exception in thread django-main-thread:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/threading.py", line 1045, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.11/threading.py", line 982, in run
    self._target(*self._args, **self._kwargs)
  File "/home/django/.venv/lib/python3.11/site-packages/django/utils/autoreload.py", line 64, in wrapper
    fn(*args, **kwargs)
  File "/home/django/.venv/lib/python3.11/site-packages/django/core/management/commands/runserver.py", line 133, in inner_run
    self.check(display_num_errors=True)
  File "/home/django/.venv/lib/python3.11/site-packages/django/core/management/base.py", line 485, in check
    all_issues = checks.run_checks(
                 ^^^^^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/core/checks/registry.py", line 88, in run_checks
    new_errors = check(app_configs=app_configs, databases=databases)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/core/checks/urls.py", line 14, in check_url_config
    return check_resolver(resolver)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/core/checks/urls.py", line 24, in check_resolver
    return check_method()
           ^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/urls/resolvers.py", line 494, in check
    for pattern in self.url_patterns:
                   ^^^^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/utils/functional.py", line 57, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
                                         ^^^^^^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/urls/resolvers.py", line 715, in url_patterns
    patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
                       ^^^^^^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/utils/functional.py", line 57, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
                                         ^^^^^^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/urls/resolvers.py", line 708, in urlconf_module
    return import_module(self.urlconf_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/home/django/.venv/lib/python3.11/site-packages/ddtrace/internal/module.py", line 220, in _exec_module
    self.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/django/app/partner_library/urls.py", line 2, in <module>
    from partner_library.graphql_schema import schema
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/home/django/.venv/lib/python3.11/site-packages/ddtrace/internal/module.py", line 220, in _exec_module
    self.loader.exec_module(module)
  File "/home/django/app/partner_library/graphql_schema.py", line 1689, in <module>
    schema = build_schema(
             ^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_federation/schema.py", line 120, in build_schema
    _add_sharable_to_page_info_type(
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_federation/schema.py", line 56, in _add_sharable_to_page_info_type
    directive_decorator(target_directive=sharable)(field=None)(PageInfo)
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_directives/directive.py", line 177, in decorator
    set_attribute_value(
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_directives/utils.py", line 64, in set_attribute_value
    raise DirectiveValidationError(
graphene_directives.exceptions.DirectiveValidationError: @shareable is got duplicate values {}, at: PageInfo
```

</details>

Maybe `_add_sharable_to_page_info_type()` should check to make sure the directive hasn't already been applied?

File: graphene_federation/apollo_versions/v2_2.py
Comment (mak626) [reply graphene_federation/apollo_versions/v2_2.py]: Thanks , have fixed this

File: graphene_federation/schema.py
Comment (mak626) [reply graphene_federation/schema.py]: Thanks. I have written a fix for this https://github.com/strollby/graphene-federation/commit/3e4013bdd13be0c7552e62857b7c44b23264583d. Let me know if it works in django

Comment (noelleleigh) [reply graphene_federation/schema.py]: New error from this change, looks like `directive_decorator()()(type_)` expects a class, not an instance:

```
File "/home/django/.venv/lib/python3.11/site-packages/graphene_federation/schema.py", line 121, in build_schema
    _add_sharable_to_page_info_type(
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_federation/schema.py", line 57, in _add_sharable_to_page_info_type
    directive_decorator(target_directive=sharable)(field=None)(page_info)
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_directives/directive.py", line 169, in decorator
    if not any(
           ^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_directives/directive.py", line 170, in <genexpr>
    LOCATION_NON_FIELD_VALIDATOR[valid_type](type_)
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_directives/constants.py", line 9, in <lambda>
    GrapheneDirectiveLocation.OBJECT: lambda t: issubclass(t, graphene.ObjectType),
                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: issubclass() arg 1 must be a class
```

Comment (mak626) [reply graphene_federation/schema.py]: Have fixed it now https://github.com/strollby/graphene-federation/commit/9fc339112be4248ea2a59d3902f88b06d6699929.  Can you recheck on django

Comment (noelleleigh) [reply graphene_federation/schema.py]: That fixed this issue, thanks!

Comment (mak626) [reply graphene_federation/schema.py]: Awesome.

**Review by noelleleigh — APPROVED** [review]
I don't have merge permissions on this repository, but GitHub seems to think I need to approve this PR since I commented on it, and I don't want to hold up this PR any more than it needs to be. I haven't thoroughly tested this.

**Pull request closed without merging**
